### PR TITLE
JIT: Replace `BlockSet` with `BitVec`, and remove BB epoch

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Part of #107749. Prerequisite for #110026.
* Use postorder number-based BitVecs in RBO and block layout.
* Use bbID-based BitVecs in `fgIncorporateProfileData`. This runs early enough in the JIT frontend such that I would expect bbIDs and bbNums to be 1:1, so I don't expect any TP impact from this change.

Switch descriptor creation still uses `bbNums` as a key into a `BitVec` as a workaround for BB epoch invariants -- I'll try switching this over to `bbID` in a follow-up to evaluate the TP cost of a sparser bitset.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib, @jakobbotsch PTAL. [No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=875271&view=ms.vss-build-web.run-extensions-tab). Thanks!

**Review by jakobbotsch — COMMENTED** [review]
Can we delete `BlockSet`/and the epoch fields as well now?

File: src/coreclr/jit/block.cpp [hunk 1]
```diff
@@ -617,22 +617,23 @@ void BasicBlock::dspSuccs(Compiler* compiler)
     // compute it ourselves here.
     if (bbKind == BBJ_SWITCH)
     {
-        // Create a set with all the successors. Don't use BlockSet, so we don't need to worry
-        // about the BlockSet epoch.
-        unsigned     bbNumMax = compiler->fgBBNumMax;
-        BitVecTraits bitVecTraits(bbNumMax + 1, compiler);
-        BitVec       uniqueSuccBlocks(BitVecOps::MakeEmpty(&bitVecTraits));
+        unsigned* const targetNums  = new unsigned[NumSucc()];
```
Comment (jakobbotsch) [inline src/coreclr/jit/block.cpp:620]:
```suggestion
        unsigned* const targetNums  = new (compiler, CMK_DebugOnly) unsigned[NumSucc()];
```

But is this change even necessary?

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 2]
```diff
@@ -3917,28 +3917,27 @@ void Compiler::fgDebugCheckBlockLinks()
...
-                // are in the BlockSet.
+                // are in the BitVec.
                 unsigned count = BitVecOps::Count(&bitVecTraits, succBlocks);
                 assert(uniqueSuccSet.numDistinctSuccs == count);
                 for (unsigned i = 0; i < uniqueSuccSet.numDistinctSuccs; i++)
                 {
                     assert(BitVecOps::IsMember(&bitVecTraits, succBlocks,
-                                               uniqueSuccSet.nonDuplicates[i]->getDestinationBlock()->bbNum));
+                                               uniqueSuccSet.nonDuplicates[i]->getDestinationBlock()->bbID));
                 }
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgdiagnostic.cpp:3941]: Why this change?

File: src/coreclr/jit/fgflow.cpp [hunk 3]
```diff
@@ -556,11 +552,11 @@ Compiler::SwitchUniqueSuccSet Compiler::GetDescriptorForSwitch(BasicBlock* switc
...
             FlowEdge* const   succEdge = swtDesc->bbsDstTab[i];
             BasicBlock* const targ     = succEdge->getDestinationBlock();
-            if (BitVecOps::IsMember(&blockVecTraits, uniqueSuccBlocks, targ->bbNum))
+            if (BitVecOps::IsMember(&blockVecTraits, uniqueSuccBlocks, targ->bbID))
             {
                 nonDups[nonDupInd] = succEdge;
                 nonDupInd++;
-                BitVecOps::RemoveElemD(&blockVecTraits, uniqueSuccBlocks, targ->bbNum);
+                BitVecOps::RemoveElemD(&blockVecTraits, uniqueSuccBlocks, targ->bbID);
             }
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgflow.cpp:560]: Ditto here, it seems unnecessary to change this.

File: src/coreclr/jit/fgdiagnostic.cpp (hunk 2 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/fgdiagnostic.cpp:3941]: It's not necessary for removing `BlockSet`, but these are the last few usages of `bbNum`-based BitVecs, so I thought I'd get rid of them while I'm here.

File: src/coreclr/jit/block.cpp (hunk 1 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/block.cpp:620]: I guess not -- if we get to the point where we don't need `bbNum` at all, we can rewrite this logic altogether to print in `bbID` order (or some other order).

File: src/coreclr/jit/fgflow.cpp (hunk 3 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/fgflow.cpp:560]: This switch to `bbID`-based `BitVecs` isn't Debug-only like the other ones, though the sparser set doesn't seem to have impacted [TP](https://dev.azure.com/dnceng-public/public/_build/results?buildId=875722&view=ms.vss-build-web.run-extensions-tab) all that much.

Comment (jakobbotsch) [reply src/coreclr/jit/fgflow.cpp]: Yeah, I would be surprised if the sparser bit sets are going to have any noticeable impact on performance.. A few bits wasted due to optimizing out entire basic blocks is going to be a drop in the bucket compared to everything else that's happening when handling the IR in those basic blocks.

My feedback here probably more comes down to bikeshedding over the naming of the final field we are going to end up with...

**Review by jakobbotsch — APPROVED** [review]
LGTM!
